include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp

```diff
@@ -58,30 +62,53 @@ namespace nil {
                         throw std::invalid_argument("expected n == (1u << logn)");
 
                     /* swapping in place (from Storer's book) */
-                    for (std::size_t k = 0; k < n; ++k) {
-                        const std::size_t rk = bitreverse(k, logn);
-                        if (k < rk)
-                            std::swap(a[k], a[rk]);
-                    }
+                    wait_for_all(ThreadPool::get_instance(0).block_execution<void>(
+                        n,
+                        [logn, &a](std::size_t begin, std::size_t end) {
+                            for (std::size_t k = begin; k < end; k++) {
+                                const std::size_t rk = crypto3::math::detail::bitreverse(k, logn);
+                                if (k < rk)
+                                    std::swap(a[k], a[rk]);
+                            }
+                        }
+                    ));
 
                     std::size_t m = 1;    // invariant: m = 2^{s-1}
+                    field_value_type w_m;
+
                     for (std::size_t s = 1; s <= logn; ++s) {
                         // w_m is 2^s-th root of unity now
-                        const field_value_type w_m = omega.pow(n / (2 * m));
-
-                        asm volatile("/* pre-inner */");
-                        for (std::size_t k = 0; k < n; k += 2 * m) {
-                            field_value_type w = field_value_type::one();
-                            for (std::size_t j = 0; j < m; ++j) {
-                                const value_type t = a[k + j + m] * w;
-                                a[k + j + m] = a[k + j] - t;
-                                a[k + j] = a[k + j] + t;
-                                w *= w_m;
+                        w_m = omega.pow(n / (2 * m));
+                        size_t count_k = n / (2 * m) + (n % (2 * m) ? 1 : 0);
+
+                        // Here we can parallelize on the both cycles with 'k' and 'm', because for each value of k and m
+                        // the ranges of array 'a' used do not intersect. Think of these 2 cycles as 1.
```

Comment:
"Cycle" usually refers to something like CPU cycles, so to avoid confusion, let's call these "loops".

Reply:
Done.